File: ios/SuiteBDE/Repositories/Application/TokenRepository.swift
//
//  TokenRepository.swift
//  BDE
//
//  Created by Nathan Fallet on 19/01/2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import Foundation
import Keychain
import shared

class TokenRepository: ITokenRepository {

    private let keychain = {
        let dev = Bundle.main.bundleIdentifier?.contains(".dev") == true
        return Keychain(accessGroup: dev ? "group.me.nathanfallet.suitebde.dev" : "group.me.nathanfallet.suitebde")
    }()

    func getToken() -> String? {
        keychain.value(forKey: "token") as? String
    }

    func getRefreshToken() -> String? {
        keychain.value(forKey: "refreshToken") as? String
    }

    func getUserId() -> CoreUUID? {
        if let uuid = keychain.value(forKey: "userId") as? UUID {
            CoreUUID(nsUUID: uuid)
        } else {
            nil
        }
    }

    func getAssociationId() -> CoreUUID? {
        if let uuid = keychain.value(forKey: "associationId") as? UUID {
            CoreUUID(nsUUID: uuid)
        } else {
            nil
        }
    }

    func getFcmToken() -> String? {
        keychain.value(forKey: "fcmToken") as? String
    }

    func setToken(token: String?) {
        let _ = if let token {
            keychain.save(token, forKey: "token")
        } else {
            keychain.remove(forKey: "token")
        }
    }

    func setRefreshToken(token: String?) {
        let _ = if let token {
            keychain.save(token, forKey: "refreshToken")
        } else {
            keychain.remove(forKey: "refreshToken")
        }
    }

    func setUserId(userId: CoreUUID?) {
        let _ = if let userId {
            keychain.save(userId.nsUUID, forKey: "userId")
        } else {
            keychain.remove(forKey: "userId")
        }
    }

    func setAssociationId(associationId: CoreUUID?) {
        let _ = if let associationId {
            keychain.save(associationId.nsUUID, forKey: "associationId")
        } else {
            keychain.remove(forKey: "associationId")
        }
    }

    func setFcmToken(fcmToken: String?) {
        let _ = if let fcmToken {
            keychain.save(fcmToken, forKey: "fcmToken")
        } else {
            keychain.remove(forKey: "fcmToken")
        }
    }

}

File: android/src/main/kotlin/com/suitebde/repositories/application/TokenRepository.kt
package com.suitebde.repositories.application

import android.content.Context
import dev.kaccelero.models.UUID

class TokenRepository(
    context: Context,
) : ITokenRepository {

    private val sharedPreferences = context.getSharedPreferences("suitebde", Context.MODE_PRIVATE)

    override fun getToken(): String? = sharedPreferences.getString("token", null)

    override fun setToken(token: String?) = sharedPreferences.edit()
        .putString("token", token)
        .apply()

    override fun getRefreshToken(): String? = sharedPreferences.getString("refreshToken", null)

    override fun setRefreshToken(token: String?) = sharedPreferences.edit()
        .putString("refreshToken", token)
        .apply()

    override fun getUserId(): UUID? = sharedPreferences.getString("userId", null)?.let(::UUID)

    override fun setUserId(userId: UUID?) = sharedPreferences.edit()
        .putString("userId", userId.toString())
        .apply()

    override fun getAssociationId(): UUID? = sharedPreferences.getString("associationId", null)?.let(::UUID)

    override fun setAssociationId(associationId: UUID?) = sharedPreferences.edit()
        .putString("associationId", associationId.toString())
        .apply()

    override fun getFcmToken(): String? = sharedPreferences.getString("fcmToken", null)

    override fun setFcmToken(fcmToken: String?) = sharedPreferences.edit()
        .putString("fcmToken", fcmToken)
        .apply()

}
